openmc/mgxs/mgxs.py

```diff
@@ -2829,6 +2832,294 @@ def nu(self, nu):
+                                          filter_bins=[('P1',)],
+                                          squeeze=True)
+            p1_tally._scores = ['scatter-1']
+
+            # Compute total cross section
+            total_xs = self.tallies['total'] / self.tallies['flux (tracklength)']
+
+            # Compute transport correction term
+            trans_corr = self.tallies['scatter-1'] / self.tallies['flux (analog)']
+
+            # Compute the diffusion coefficient
+            transport = total_xs - trans_corr
+            dif_coef = transport**(-1) / 3.0
+
+            self._rxn_rate_tally = dif_coef * self.tallies['flux (tracklength)']
+            self._rxn_rate_tally.sparse = self.sparse
+
+        return self._rxn_rate_tally
+
+    @property
+    def xs_tally(self):
+        if self._xs_tally is None:
+            if self.tallies is None:
+                msg = 'Unable to get xs_tally since tallies have ' \
+                      'not been loaded from a statepoint'
+                raise ValueError(msg)
+
+            self._xs_tally = self.rxn_rate_tally / self.tallies['flux (tracklength)']
+            self._compute_xs()
+
+        return self._xs_tally
+
+    def get_condensed_xs(self, coarse_groups, condense_dif_coef=True):
+        """Construct an energy-condensed version of this cross section.
+        Parameters
+        ----------
+        coarse_groups : openmc.mgxs.EnergyGroups
+            The coarse energy group structure of interest
+        Returns
+        -------
+        MGXS
+            A new MGXS condensed to the group structure of interest
+        """
+
+        cv.check_type('coarse_groups', coarse_groups, EnergyGroups)
+        cv.check_less_than('coarse groups', coarse_groups.num_groups,
+                           self.num_groups, equality=True)
+        cv.check_value('upper coarse energy', coarse_groups.group_edges[-1],
+                       [self.energy_groups.group_edges[-1]])
+        cv.check_value('lower coarse energy', coarse_groups.group_edges[0],
+                       [self.energy_groups.group_edges[0]])
+
+        # Clone this MGXS to initialize the condensed version
+        condensed_xs = copy.deepcopy(self)
+
+        if condense_dif_coef:
+            if self._rxn_rate_tally is None:
+
+                p1_tally = self.tallies['scatter-1']
+                old_filt = p1_tally.filters[-2]
+                new_filt = openmc.EnergyFilter(old_filt.values)
+                p1_tally.filters[-2] = new_filt
+
+                # Slice Legendre expansion filter and change name of score
+                p1_tally = p1_tally.get_slice(filters=[openmc.LegendreFilter],
+                        filter_bins=[('P1',)],
+                        squeeze=True)
+                p1_tally._scores = ['scatter-1']
+
+            total = self.tallies['total'] / self.tallies['flux (tracklength)']
+            trans_corr = p1_tally / self.tallies['flux (analog)']
```

Comment:
If self._rxn_rate_tally is not None, then where would p1_tally come from?

Reply:
Good point. I could just remove the statement "if self._rxn_rate_tally is None".

Reply:
Actually, can you explain this question to me some more? I see the same structure in TransportXS and am wondering how it is resolved in that class.

Reply:
That structure is in ``TransportXS.rxn_rate_tally`` (I think that's what you're referring to), but in that case it acts only as a means to not re-calculate something that has already been computed as it can be relatively costly.

Here, I'm not sure what the purpose of the ``if self._rxn_rate_tally is None`` block is. If you compare this block with what is in ``def rxn_rate_tally`` for the Diffusion Coefficient, it looks like you can replace
```
            if self._rxn_rate_tally is None:

                p1_tally = self.tallies['scatter-1']
                old_filt = p1_tally.filters[-2]
                new_filt = openmc.EnergyFilter(old_filt.values)
                p1_tally.filters[-2] = new_filt

                # Slice Legendre expansion filter and change name of score
                p1_tally = p1_tally.get_slice(filters=[openmc.LegendreFilter],
                        filter_bins=[('P1',)],
                        squeeze=True)
                p1_tally._scores = ['scatter-1']

            total = self.tallies['total'] / self.tallies['flux (tracklength)']
            trans_corr = p1_tally / self.tallies['flux (analog)']
            transport = (total - trans_corr)
            dif_coef = transport**(-1) / 3.0
            dif_coef *= self.tallies['flux (tracklength)']
```

with one of the two options:

```
            if self._rxn_rate_tally is None:

                p1_tally = self.tallies['scatter-1']
                old_filt = p1_tally.filters[-2]
                new_filt = openmc.EnergyFilter(old_filt.values)
                p1_tally.filters[-2] = new_filt

                # Slice Legendre expansion filter and change name of score
                p1_tally = p1_tally.get_slice(filters=[openmc.LegendreFilter],
                        filter_bins=[('P1',)],
                        squeeze=True)
                p1_tally._scores = ['scatter-1']

                total = self.tallies['total'] / self.tallies['flux (tracklength)']
                trans_corr = p1_tally / self.tallies['flux (analog)']
                transport = (total - trans_corr)
                dif_coef = transport**(-1) / 3.0
                dif_coef *= self.tallies['flux (tracklength)']
```

OR

```
            dif_coef = self.rxn_rate_tally
            <insert code to convet dif_coef from sparse to dense as it looks like the rest of the function expects dense>
```

Reply:
Real quick on this one: you added in the if/else block here; if ``dif_coef = self.rxn_rate_tally`` works fine, then cant you just drop the whole if block and only use ``dif_coef = self.rxn_rate_tally``? 

(the ``rxn_rate_tally`` property includes the ``if self._rxn_rate_tally is None`` block already.

Reply:
I did try that, but the results were different and resulted in mgxs_library_condense failing. 

When I only use dif_coef = self.rxn_rate_tally, it doesn't slice out the Legendre expansion. A series of print-statements showed me that it definitely calls the property self.rxn_rate_tally(self) which has all the same code. But it didn't do the slice for some reason.

When I used the if/else block, the slicing occurred as expected and the test passed.